src/System.Threading.AccessControl/src/System/Threading/EventWaitHandleAcl.cs

```diff
@@ -0,0 +1,93 @@
+        /// <param name="eventSecurity">The Windows access control security to apply.</param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentException">The length of the name exceeds the maximum limit.</exception>
+        /// <exception cref="ArgumentNullException"><paramref name="eventSecurity" /> is <see langword="null" />.</exception>
+        /// <exception cref="ArgumentOutOfRangeException">The <paramref name="mode" /> enum value was out of legal range.</exception>
+        /// <exception cref="DirectoryNotFoundException">Could not find a part of the path specified in <paramref name="name" />.</exception>
+        /// <exception cref="WaitHandleCannotBeOpenedException">An event with the provided <paramref name="name" /> was not found.
+        /// -or-
+        /// An <see cref="EventWaitHandle" /> with system-wide name <paramref name="name" /> cannot be created. An <see cref="EventWaitHandle" /> of a different type might have the same name.</exception>
+        public static unsafe EventWaitHandle Create(bool initialState, EventResetMode mode, string name, out bool createdNew, EventWaitHandleSecurity eventSecurity)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentException(SR.Format(SR.Argument_CannotBeNullOrEmpty), nameof(name));
+            }
+            if (name.Length > Interop.Kernel32.MAX_PATH)
+            {
+                throw new ArgumentException(SR.Format(SR.Argument_WaitHandleNameTooLong, name), nameof(name));
+            }
+            if (eventSecurity == null)
+            {
+                throw new ArgumentNullException(nameof(eventSecurity));
+            }
+
+            bool isManualReset = mode switch
+            {
+                EventResetMode.ManualReset => true,
+                EventResetMode.AutoReset => false,
+                _ => throw new ArgumentOutOfRangeException(nameof(mode), SR.Format(SR.ArgumentOutOfRange_Enum))
+            };
+
+            fixed (byte* pSecurityDescriptor = eventSecurity.GetSecurityDescriptorBinaryForm())
+            {
+                var secAttrs = new Interop.Kernel32.SECURITY_ATTRIBUTES
+                {
+                    nLength = (uint)sizeof(Interop.Kernel32.SECURITY_ATTRIBUTES),
+                    lpSecurityDescriptor = (IntPtr)pSecurityDescriptor
+                };
+
+                using SafeWaitHandle handle = Interop.Kernel32.CreateEvent(ref secAttrs, isManualReset, initialState, name);
+
+                ValidateHandle(handle, name, out createdNew);
+
+                try
+                {
+                    return EventWaitHandle.OpenExisting(name);
+                }
+                finally
+                {
+                    // Close our handle as the EventWaitHandle will have it's own.
+                    handle.Dispose();
+                }
```

Comment:
I continue to not understand why we're placing the `name != null` restriction.  Why can't this instead be, e.g.
```C#
SafeWaitHandle handle = Interop.Kernel32.CreateEvent(ref secAttrs, isManualReset, initialState, name);
ValidateHandle(handle, name, out createdNew);

var ewh = new EventWaitHandle(initialState, mode);
var old = ewh.SafeWaitHandle;
ewh.SafeWaitHandle = handle;
old.Dispose();
```
?  Does that not work for some reason?

Reply:
It does not work.
Unit tests that are passing with my current code, start failing with callstacks like the one below when I add your suggested change.
The exception shows up when attempting to retrieve the access control from the newly created EventWaitHandle.
For example:
```
System.Threading.Tests.ThreadingAclExtensionsTests.EventWaitHandle_Create_SpecificParameters(initialState: True, mode: AutoReset, accessControl: Allow, rights: FullControl) [FAIL]

System.ObjectDisposedException : Safe handle has been closed.

Object name: 'SafeHandle'.

Stack Trace:
/_/src/System.Private.CoreLib/shared/System/Runtime/InteropServices/SafeHandle.cs(156,0): at System.Runtime.InteropServices.SafeHandle.DangerousAddRef(Boolean& success)
/_/src/System.Private.CoreLib/src/System/StubHelpers.cs(1747,0): at System.StubHelpers.StubHelpers.SafeHandleAddRef(SafeHandle pHandle, Boolean& success)
at Interop.Advapi32.GetSecurityInfoByHandle(SafeHandle handle, UInt32 objectType, UInt32 securityInformation, IntPtr& sidOwner, IntPtr& sidGroup, IntPtr& dacl, IntPtr& sacl, IntPtr& securityDescriptor)
D:\corefx\src\System.Security.AccessControl\src\System\Security\AccessControl\Win32.cs(134,0): at System.Security.AccessControl.Win32.GetSecurityInfo(ResourceType resourceType, String name, SafeHandle handle, AccessControlSections accessControlSections, RawSecur
ityDescriptor& resultSd)
D:\corefx\src\System.Security.AccessControl\src\System\Security\AccessControl\NativeObjectSecurity.cs(111,0): at System.Security.AccessControl.NativeObjectSecurity.CreateInternal(ResourceType resourceType, Boolean isContainer, String name, SafeHandle handle, Acc
essControlSections includeSections, Boolean createByName, ExceptionFromErrorCode exceptionFromErrorCode, Object exceptionContext)
D:\corefx\src\System.Security.AccessControl\src\System\Security\AccessControl\NativeObjectSecurity.cs(84,0): at System.Security.AccessControl.NativeObjectSecurity..ctor(Boolean isContainer, ResourceType resourceType, SafeHandle handle, AccessControlSections incl
udeSections, ExceptionFromErrorCode exceptionFromErrorCode, Object exceptionContext)
D:\corefx\src\System.Threading.AccessControl\src\System\Security\AccessControl\EventWaitHandleSecurity.cs(125,0): at System.Security.AccessControl.EventWaitHandleSecurity..ctor(SafeWaitHandle handle, AccessControlSections includeSections)
D:\corefx\src\System.Threading.AccessControl\src\System\Threading\ThreadingAclExtensions.cs(17,0): at System.Threading.ThreadingAclExtensions.GetAccessControl(EventWaitHandle handle)
D:\corefx\src\System.Threading.AccessControl\tests\ThreadingAclExtensionsTests.cs(171,0): at System.Threading.Tests.ThreadingAclExtensionsTests.CreateAndVerifyEventWaitHandle(Boolean initialState, EventResetMode mode, EventWaitHandleSecurity expectedSecurity)
D:\corefx\src\System.Threading.AccessControl\tests\ThreadingAclExtensionsTests.cs(124,0): at System.Threading.Tests.ThreadingAclExtensionsTests.EventWaitHandle_Create_SpecificParameters(Boolean initialState, EventResetMode mode, AccessControlType accessControl,
EventWaitHandleRights rights)
```

Reply:
> It does not work.

Please share the code you used.

Reply:
> I continue to not understand why we're placing the name != null restriction.

In theory you could, it would be creating and tossing a global object, of course. I think I'd rather just "fake" it and create and modify if we're going to support it. In theory there shouldn't be risk here as you need the handle to the WaitHandle to do anything with it when it is unnamed. If the security is modified before we give it back it should be good...

@carlossanlop can you please share a more specific repro example or explain why it is failing- afaik, it shouldn't.

Reply:
> _Please share the code you used._

> _can you please share a more specific repro example or explain why it is failing- afaik, it shouldn't._

@stephentoub @JeremyKuhne yes of course, gladly. I submitted a new commit with the applied suggestion so the CI hits the failure I mentioned. Please take a look.

Reply:
> Please take a look.

You didn't remove the `using`... that's why you're getting an ObjectDisposedException.

Reply:
Yes, thanks for catching it. Now it works.